File: 18-4sum/18-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
      List<List<Integer>> ans = new LinkedList<>();
      for(int i = 0; i < nums.length; i++){
          if(i > 0 && nums[i] == nums[i-1]){
              continue;
          }else{
              for(int j = i+1; j < nums.length; j++){
                  if(j > i+1 && nums[j] == nums[j-1]){
                      continue;
                  }else{
                      var l = j+1;
                      var r = nums.length-1;
                      while(l < nums.length &&  r > 0 && l < r){
                          long sum = (long)nums[i]+ (long)nums[j]+ (long)nums[l]+ (long)nums[r];
                          if(sum == target){
                              ans.add(List.of(nums[i], nums[j], nums[l], nums[r]));
                              l++;
                              while(l < nums.length && nums[l] == nums[l-1] && l < r){l++;}
                          }else if(sum < target){
                              l++;
                          }else{
                              r--;
                          }
                      }
                  }
              }
          }
      }
      return ans;
    }
}

File: 18-4sum/18-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
          vector<vector<int>> res;
        
        sort(nums.begin(),nums.end());
        
        int n=nums.size();
        
        
        
        for(int i=0;i<n;i++)
        {   
            if(i!=0 && nums[i]==nums[i-1])
               continue;
            for(int j=i+1;j<n;j++)
            {
                //4 sum to 3 sum
                if(j!=i+1 && nums[j]==nums[j-1])
                        continue;
                
                  long long tgt=(long long)target-(long long)nums[i] -(long long)nums[j];
                
                
                 int fst=j+1;
                int scnd=n-1;
            
                if(fst>=scnd)
                    continue;
            
            while(fst<scnd)
            {
                if((long long)(nums[fst]+nums[scnd])==tgt)
                {
                    res.push_back({nums[fst],nums[scnd],nums[i],nums[j]});
                    
                    fst++;
                   while(fst<scnd && nums[fst]==nums[fst-1])
                       fst++;
                    scnd--;
                    while(scnd>fst && nums[scnd]==nums[scnd+1])
                       scnd--;
                    
                    
                }
                
                
                else if((long long)(nums[fst]+nums[scnd])>tgt)
                    scnd--;
                else
                    fst++;
            
            
            }
                
                
            }
        }
      return res;       

    }
};